components/server/src/workspace/ec2-workspace-manager.ts

```diff
@@ -0,0 +1,724 @@
+import { TokenProvider } from "../user/token-provider";
+import { MessageBusIntegration } from "./messagebus-integration";
+import { URL } from 'url';
+import * as ec2 from '@aws-sdk/client-ec2';
+import * as ssm from '@aws-sdk/client-ssm';
+import * as http from 'http';
+import * as https from 'https';
+
+interface EC2WorkspaceInstance {
+    workspaceId?: string;
+    ec2InstanceId?: string;
+};
+
+const enum EC2InstanceStatus {
+    UNDEFINED = -1,
+    PENDING = 0,
+    RUNNING = 16,
+    SHUTTING_DOWN = 32,
+    TERMINATED = 48,
+    STOPPING = 64,
+    STOPPED = 80
+};
+
+interface EC2WorkspaceDetails {
+    workspace: EC2WorkspaceInstance;
+    ec2StatusCode: EC2InstanceStatus;
+    ec2SystemStatus: boolean;
+    ec2InstanceStatus: boolean;
+};
+
+@injectable()
+export class EC2WorkspaceManager {
+    @inject(Env) protected readonly env: Env;
+    @inject(TracedWorkspaceDB) protected readonly workspaceDb: DBWithTracing<WorkspaceDB>;
+    @inject(TokenProvider) protected readonly tokenProvider: TokenProvider;
+    @inject(HostContextProvider) protected readonly hostContextProvider: HostContextProvider;
+    @inject(MessageBusIntegration) protected readonly messageBus: MessageBusIntegration;
+    @inject(SessionHandlerProvider) protected sessionHandlerProvider: SessionHandlerProvider;
+
+    // Path to store all devspaces configuration files on the instance
+    private readonly workspaceConfigDir = '~/.devspaces';
+
+    public isEC2WorkspaceInstance(instance: WorkspaceInstance): boolean {
+        const raw_regex_pattern = '^https://windows-[0-9]{1,3}-[0-9]{1,3}-[0-9]{1,3}-[0-9]{1,3}\\.ws\\.' + this.env.gitpodDomain.replace(/\./gi, "\\.");
+        const ec2_regex = new RegExp(raw_regex_pattern);
+        return ec2_regex.test(instance.ideUrl);
+    }
+
+    public async isEC2WorkspaceId(workspaceId: string): Promise<boolean> {
+        const desc = await this.describeWorkspace(this.newEC2Client(), { workspaceId: workspaceId });
+        const statusCode: number = desc?.ec2StatusCode || EC2InstanceStatus.UNDEFINED;
+        switch (statusCode) {
+            case EC2InstanceStatus.PENDING:
+            case EC2InstanceStatus.RUNNING:
+            case EC2InstanceStatus.STOPPING:
+            case EC2InstanceStatus.STOPPED:
+                // Usable workspace, can be deleted
+                return true;
+            default:
+                // Absent or already terminated
+                return false;
+        }
+    }
+
+    public async delay(ms: number) {
+        return new Promise( resolve => setTimeout(resolve, ms) );
+    }
+
+    private getWorkspaceInstanceAddress(instance: WorkspaceInstance): string | undefined {
+        // TODO: When this URL pattern is deprecated, we shall use describeWorkspace instead.
+        const raw_regex_pattern = '^https://windows-(?<IPAddress>[0-9]{1,3}-[0-9]{1,3}-[0-9]{1,3}-[0-9]{1,3})\\.ws\\.' + this.env.gitpodDomain.replace(/\./gi, "\\.");
```

Comment:
all pattern code to detect windows needs to be move a common function, so that we can fix or enhance in one place.

Reply:
I'll refactor this. This regex has a capture group and the other one doesn't. I'll see if I can merge the two before merging this.